spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java

```diff
@@ -869,15 +871,9 @@ public Message receive(long timeoutMillis) throws AmqpException {
 	@Override
 	public Message receive(final String queueName, final long timeoutMillis) {
 		Message message = execute(channel -> {
-			QueueingConsumer consumer = createQueueingConsumer(queueName, channel);
-			Delivery delivery;
-			if (timeoutMillis < 0) {
-				delivery = consumer.nextDelivery();
-			}
-			else {
-				delivery = consumer.nextDelivery(timeoutMillis);
-			}
-			channel.basicCancel(consumer.getConsumerTag());
+			CompletableFuture<Delivery> future = new CompletableFuture<>();
+			Delivery delivery = getDeliveryAndCancelConsumer(channel, future,
+					createConsumer(queueName, channel, future), timeoutMillis);
```

Comment:
Looks like `future` variable and `createConsumer()` logic can be hide in the `getDeliveryAndCancelConsumer()`.